torch/optim/optimizer.py

```diff
@@ -127,12 +127,36 @@ def _disable_dynamo_if_unsupported(single_tensor_fn=None):
         globals()[single_tensor_fn.__name__] = single_tensor_fn
 
     def wrapper(func):
+        import inspect
+
+        disabled_func = torch._disable_dynamo(func)
+        ps = inspect.signature(func).parameters
+        has_state_steps = True
+        try:
+            state_steps_ind = list(ps.keys()).index("state_steps")
+        except ValueError:
+            has_state_steps = False
+
+        # Today, there are cases where we stack state steps
+        # and pass them as the value arg of foreach ops.
+        # Having state steps on cuda as the value arg is not supported in eager,
+        # but this only occurs in the rare case that the user explicitly deletes
+        # the capturable flag. If capturable=True, this is not a problem.
         @functools.wraps(func)
-        def maybe_fallback(self, *args, **kwargs):
-            if is_compiling() and not kwargs.get("capturable", False):
-                return torch._disable_dynamo(func(self, *args, **kwargs))
+        def maybe_fallback(*args, **kwargs):
```

Comment:
From my understanding of our discussion, the only two times where compile should continue compiling is if capturable=True or if params are on CPU. So the conditional I am expecting is:

if not capturable and not all params on CPU, then disable.

It may be even better to use the flip conditional to be clear:
```
if capturable or if params are all on CPU:
   don't disable
else:
   disable
```

Reply:
We support mixed cpu/gpu (and the compiler no longer sets capturable=True in this case). Here's my perspective on our discussion - it established that 1) we don't always set capturable=True (only when safe) and 2) we will now trace the non-capturable path, avoiding item calls in the short term until the scalar translation layer is ready. 

The compiler is able to handle all of the cases eager can except if capturable=False and step is on GPU.